iceoryx_utils/include/iceoryx_utils/internal/lockfree_queue/cyclic_index.hpp

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <limits>
+#include <stdint.h>
+#include <type_traits>
+
+
+namespace iox
+{
+// @todo: unit tests - will be tested with the resize feature implementation
+// which might require minor adaptations/extensions of the CyclicIndex
+
+/// @brief index structure that can contain logical values 0, ..., CycleLength-1
+/// but also stores an internal cycle counter to be used in compare_exchange
+template <uint64_t CycleLength, typename ValueType = uint64_t>
+class CyclicIndex
+{
+  public:
+    using value_t = ValueType;

```

Comment:
What is this for ?

Reply:
This is to access the type of the template parameter. However, the naming is not ideal. If we have something like template<typename ValueType> we would sometimes like to use the type ValueType in other constructions that depend on it. But we cannot without a using directive, a shortcoming of the language in my opinion. But we also cannot have using ValueType = ValueType, we need a different name. What should this name be? We need a convention here I guess.